src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -28,24 +28,26 @@ public sealed partial class Pipe
         private static readonly SendOrPostCallback s_syncContextExecuteWithoutExecutionContextCallback = ExecuteWithoutExecutionContext!;
         private static readonly Action<object?> s_scheduleWithExecutionContextCallback = ExecuteWithExecutionContext!;
 
-        // This sync objects protects the shared state between the writer and reader (most of this class)
-        private readonly object _sync = new object();
-
-        private readonly MemoryPool<byte>? _pool;
-        private readonly int _minimumSegmentSize;
-        private readonly long _pauseWriterThreshold;
-        private readonly long _resumeWriterThreshold;
-
-        private readonly PipeScheduler _readerScheduler;
-        private readonly PipeScheduler _writerScheduler;
-
         // Mutable struct! Don't make this readonly
         private BufferSegmentStack _bufferSegmentPool;
 
         private readonly DefaultPipeReader _reader;
         private readonly DefaultPipeWriter _writer;
 
-        private readonly bool _useSynchronizationContext;
+        // The options instance
+        private readonly PipeOptions _options;
+
+        // Computed state from the options instance
+        private bool UseSynchronizationContext => _options.UseSynchronizationContext;
+        private int MinimumSegmentSize => _options.MinimumSegmentSize;
+        private long PauseWriterThreshold => _options.PauseWriterThreshold;
+        private long ResumeWriterThreshold => _options.ResumeWriterThreshold;
+
+        private PipeScheduler ReaderScheduler => _options.ReaderScheduler;
+        private PipeScheduler WriterScheduler => _options.WriterScheduler;
+
+        // This sync objects protects the shared state between the writer and reader (most of this class)
+        private object SyncObj => this;
```

Comment:
Isn't this pattern frowned upon? Anyone can lock the Pipe and potentially create a deadlock (probably an uncommon scenario, but still)

Reply:
It is. It's a tradeoff of course. Since this type is exposed public there's a risk that it could cause a deadlock but I think the space saving is worth it and locking the pipe causing a deadlock seems unlikely... Though we know customers do crazy things.

@AArnott how do you feel about this?

Reply:
I think the upside here is that 99% of the time you're dealing with an PipeReader or PipeWriter.

Reply:
I have always doubted the value of avoiding `this` as the lock object. But I have nearly always avoided it anyway. When I have only one field to protect, I often _do_ lock that field's reference as a means to protect it. Usually this is a List or a Dictionary. Given Pipe is so crucial to avoid concurrency for, I can't deny folks may be locking the `Pipe` itself already. Hopefully when doing so, they only call into the pipe and other simple APIs as is good practice in any lock. And if they lock the pipe and then call into that pipe, so long as _that_ won't deadlock, I think it is reasonably safe.

That said, I don't know in what scenarios saving ~100 bytes on `Pipe` instances is worth. We have a handful (100s) of long-lived ones, and more recently we may end up having many very short-lived ones. If this change makes it less likely that we will need to pool and recycle them, I guess that's nice, but I don't think it'll be particularly valuable to scenarios that I'm familiar with.

Reply:
I'm tackling the large concurrent number of websockets scenario (https://github.com/dotnet/runtime/issues/49019). Pipe is large as well as the socket implementation. I'd like to reduce the per connection memory as much as possible.

Reply:
> That said, I don't know in what scenarios saving ~100 bytes on `Pipe` instances is worth. 

If you have 1M open WebSockets; probably valuable? https://twitter.com/Nick_Craver/status/1272607699948253187

Reply:
> That said, I don't know in what scenarios saving ~100 bytes on `Pipe` instances is worth. 

If you have 1M open sockets; probably valuable? https://twitter.com/Nick_Craver/status/1272607699948253187

Reply:
> When I have only one field to protect, I often do lock that field's reference as a means to protect it

Right, this is the primary concern. Devs that pay attention to allocation will often lock on an object they have and that they think is internal to their implementation; if that instance locks on itself, there are potential unintended interactions between the locking in the caller and the locking in the instance... that could mean unintentional contention leading to slowdowns, or deadlocks.  It also potentially becomes a breaking change to change it back, because the calling code could have unintentionally been relying on that synchronization for correctness.  We've hit this in a few cases elsewhere, e.g. we tried to change Console to not lock on its Out instance, but we broke code that was also locking on Console.Out and relying on it for coordination of multiple console interactions.

Reply:
True in the general case, I'm having a hard time fabricating cases where it's problematic. Usually, code is dealing with a PipeReader and PipeWriter (you have 3 fields to lock on by default, the `PipeReader` and `PipeWriter` and the `Pipe`).  I can see somebody locking on the pipe for to protect read/write operations. Short of doing something like this:

Thread 1:

var slim = new ManualResetEventSlim();
lock (pipe)
{
    slim.Wait();
}

Thread 2:

var memory = pipe.Writer.GetMemory();
slim.Set();

I shall ponder this a bit more.

Reply:
I have an idea....

Reply:
This should be resolved with the latest commit.

Reply:
We changed our collections a while back so that the sync obj became ‘this’ and no problems were reported - but of course it depends on usage patterns...

Reply:
> We changed our collections a while back so that the sync obj became ‘this’ and no problems were reported

That's different, though.  `ICollection.SyncRoot` is there to make the object to use for locking public, and it's explicitly documented that it should typically be `this`.  And there's a reason (multiple) this pattern wasn't carried forward into `ICollection<T>` :wink: